### PR TITLE
Guard the Map.map of immutable map with exists clause

### PR DESCRIPTION
**Pull request by diesalbla**
The `map` method of the immutable.Map class may allocate an entire new Map, since it cannot know in advance if the set of keys changes. Since this can be an expensive operation, we avoid it by first
checking if any entry in the Map would be updated.

### CODE REVIEW

**Comment by retronym** [comment]
This phase is certainly ripe for further optimization. I actually have some WIP you might want to take a look at: https://github.com/retronym/scala/tree/topic/spec-unzip. I'm tring to reduce the number of `.unzip` calls in the phase.

Your change looks to be independent of those changes.

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala [hunk 1]
```diff
@@ -574,17 +574,27 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
...
-    case (sym, AnyRefTpe) if sym.owner == origsym => (sym, typeParamSubAnyRef(sym, specsym))
-    case x => x
+  private def mapAnyRefsInSpecSym(env: TypeEnv, origsym: Symbol, specsym: Symbol): TypeEnv = {
+    val needsChange = env.exists {  case (key, value) => key.owner == origsym && value == AnyRefTpe }
+    if (needsChange)
+      env map {
+        case (sym, AnyRefTpe) if sym.owner == origsym => (sym, typeParamSubAnyRef(sym, specsym))
+        case x => x
+      }
+    else env
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:581]: Let's avoid this minor duplication.

```scala
def needsSubst(sym: Symbol, tpe: Type) = sym.owner == origsym && tpe == AnyRefTpe 
val needsChange = env.exists { case (sym, tpe) => needsSubst(sym, tpe)
...
  env map ( binding => if (needSubst(binding._1, binding._2) ... else binding)
```

Same for below.

**Review by retronym — DISMISSED** [review]
DRY things up a touch

**Comment by diesalbla** [comment]
I have DRYed them out to a single higuer-order method, that takes as parameter the function to compute the new value for each key (symbol).

**Comment by retronym** [comment]
Thinking about this some more, I remembered that `immutable.HashMap` has a `transform` method which we should use rather than `map` when we only need to modify the values, not the keys.

Furthermore, this method has been optimized in 2.13.x to perform structural sharing when the transform is an identity. This means we don't need to do a "read only" pass first with `.exists` to optimize for a common case of identity mapping functions.

I also noticed this optimization was not yet implemented for the hand-specialized, low arity `Map`-s.

I have fixed this and moved to using `transform` in the specialization phase in #7859, targetted to 2.13.x.

We should either mark the commit this PR as `[nomerge]`, or just close this in the knowledge that we'll get the performance benefit soon enough in the 2.13 series.

**Review by retronym — CHANGES_REQUESTED** [review]
If you agree my 2.13.x solution is okay as the longer term fix, either mark as `[nomerge]` or close

**Comment by adriaanm** [comment]
@diesalbla how do you feel about leaving this as-is in 2.12? I'm open to merging with the amended commit title as well.

**Comment by diesalbla** [comment]
@adriaanm That will be fine. I had almost forgotten about this one.

**Pull request closed without merging**